File: contests/lc/840.MagicSquaresInGrid/solution.py
from typing import List


class Solution:
    def numMagicSquaresInside(self, grid: List[List[int]]) -> int:
        R, C = len(grid), len(grid[0])

        def is_magic(r: int, c: int) -> bool:
            # quick prune: center must be 5 in 1..9 magic square
            if grid[r + 1][c + 1] != 5:
                return False

            seen = set()
            for i in range(r, r + 3):
                for j in range(c, c + 3):
                    v = grid[i][j]
                    if v < 1 or v > 9 or v in seen:
                        return False
                    seen.add(v)

            s = sum(grid[r][c:c + 3])
            if sum(grid[r + 1][c:c + 3]) != s or sum(grid[r + 2][c:c + 3]) != s:
                return False

            if (grid[r][c] + grid[r + 1][c] + grid[r + 2][c]) != s:
                return False
            if (grid[r][c + 1] + grid[r + 1][c + 1] + grid[r + 2][c + 1]) != s:
                return False
            if (grid[r][c + 2] + grid[r + 1][c + 2] + grid[r + 2][c + 2]) != s:
                return False

            if (grid[r][c] + grid[r + 1][c + 1] + grid[r + 2][c + 2]) != s:
                return False
            if (grid[r][c + 2] + grid[r + 1][c + 1] + grid[r + 2][c]) != s:
                return False

            return True

        ans = 0
        for r in range(R - 2):
            for c in range(C - 2):
                if is_magic(r, c):
                    ans += 1
        return ans

File: contests/lc/840.MagicSquaresInGrid/solution.go
package magicsquaresingrid

func isMagicSquare(grid [][]int, row, col int) bool {
	// Extract the 3x3 subgrid
	square := make([][]int, 3)
	for i := 0; i < 3; i++ {
		square[i] = make([]int, 3)
		for j := 0; j < 3; j++ {
			square[i][j] = grid[row+i][col+j]
		}
	}

	// Check if all numbers from 1 to 9 are present
	seen := make([]bool, 10)
	for i := 0; i < 3; i++ {
		for j := 0; j < 3; j++ {
			num := square[i][j]
			if num < 1 || num > 9 {
				return false
			}
			if seen[num] {
				return false
			}
			seen[num] = true
		}
	}

	// Check rows
	rowSums := make([]int, 3)
	for i := 0; i < 3; i++ {
		for j := 0; j < 3; j++ {
			rowSums[i] += square[i][j]
		}
	}

	// Check if all rows have the same sum
	targetSum := rowSums[0]
	for i := 1; i < 3; i++ {
		if rowSums[i] != targetSum {
			return false
		}
	}

	// Check columns
	colSums := make([]int, 3)
	for j := 0; j < 3; j++ {
		for i := 0; i < 3; i++ {
			colSums[j] += square[i][j]
		}
		if colSums[j] != targetSum {
			return false
		}
	}

	// Check diagonals
	diag1Sum := square[0][0] + square[1][1] + square[2][2]
	diag2Sum := square[0][2] + square[1][1] + square[2][0]
	return diag1Sum == targetSum && diag2Sum == targetSum
}

func numMagicSquaresInside(grid [][]int) int {
	rows := len(grid)
	cols := len(grid[0])
	count := 0

	// Iterate through all possible top-left positions for a 3x3 subgrid
	for row := 0; row <= rows-3; row++ {
		for col := 0; col <= cols-3; col++ {
			if isMagicSquare(grid, row, col) {
				count++
			}
		}
	}

	return count
}
